
# DL Assignment 03

**Name:**

**Course Email:**  


## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।

# General Instruction

You must choose your own dataset.

The dataset must:

Be a supervised learning dataset (Regression or Binary Classification)

Contain at least 300 samples

Have at least 2 input features

Be in CSV format

You are NOT allowed to use Dataset or DataLoader.

You must implement everything manually.

# Question 01: [ Marks 05 ]

## Dataset Preparation

## Using your chosen dataset:

Load the dataset.

Perform necessary preprocessing:

Handle missing values (if any)

Encode categorical variables (if necessary)

Feature scaling (if needed)

Separate features (X) and target (y).

Convert them into NumPy arrays.

Convert them into PyTorch tensors.

Split into training and testing sets.

Clearly explain each preprocessing decision.

# **Write** Answer 01:


In [25]:
import pandas as pd
import numpy as np
import torch
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [26]:
df= pd.read_csv('diabetes.csv')
df

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1
...,...,...,...,...,...,...,...,...,...
763,10,101,76,48,180,32.9,0.171,63,0
764,2,122,70,27,0,36.8,0.340,27,0
765,5,121,72,23,112,26.2,0.245,30,0
766,1,126,60,0,0,30.1,0.349,47,1


In [27]:
print(df.isnull().sum())

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


In [28]:
X = df.drop('Outcome', axis=1).values
y = df['Outcome'].values

In [29]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

X_np = np.array(X_scaled)
y_np = np.array(y)

X_tensor = torch.tensor(X_np, dtype=torch.float32)
y_tensor = torch.tensor(y_np, dtype=torch.float32).view(-1, 1)

X_train, X_test, y_train, y_test = train_test_split(
    X_tensor, y_tensor, test_size=0.2, random_state=42
)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: torch.Size([614, 8])
X_test shape: torch.Size([154, 8])
y_train shape: torch.Size([614, 1])
y_test shape: torch.Size([154, 1])


# Question 02: [ Marks 20 ]

## Design a neural network using nn.Module.

### The model must contain:

Input layer

At least one hidden layer

Output layer

Suitable activation function



## Justify:

Number of hidden neurons

Choice of activation function

Print  the total number of trainable parameters.


## Write Answer 02:


In [30]:
import torch
import torch.nn as nn

In [31]:
class DiabetesModel(nn.Module):
    def __init__(self):
        super(DiabetesModel, self).__init__()
        self.fc1 = nn.Linear(8, 16)

        self.relu = nn.ReLU()

        self.fc2 = nn.Linear(16, 1)

        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        x = self.sigmoid(x)
        return x

model = DiabetesModel()

print(model)

total_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print("Total Trainable Parameters:", total_params)

DiabetesModel(
  (fc1): Linear(in_features=8, out_features=16, bias=True)
  (relu): ReLU()
  (fc2): Linear(in_features=16, out_features=1, bias=True)
  (sigmoid): Sigmoid()
)
Total Trainable Parameters: 161


## 1. Input Layer

The dataset has **8 input features**, so the input layer contains:

$$
text{Input neurons} = 8
$$

**Example features:**
- Pregnancies  
- Glucose  
- BloodPressure  
- BMI  
- etc.


## 2. Hidden Layer

I used **16 hidden neurons**.

$$
text{Hidden neurons} = 16
$$

**Justification:**
- More neurons allow the model to learn complex patterns  
- 16 provides a good balance between learning capacity and avoiding overfitting  
- Suitable for small to medium datasets  

**Network structure:**
$$
8 \rightarrow 16 \rightarrow 1
$$


## 3. Output Layer

The output layer has:

$$
text{Output neurons} = 1
$$

**Reason:**
- This is a binary classification problem  
- Output classes:
  - \(0\) = No Diabetes  
  - \(1\) = Diabetes  


## 4. Choice of Activation Function

### Hidden Layer: ReLU

$$
text{ReLU}(x) = \max(0, x)
$$

**Why ReLU?**
- Computationally efficient  
- Helps reduce vanishing gradient problem  
- Faster training  


### Output Layer: Sigmoid

$$
sigma(x) = \frac{1}{1 + e^{-x}}
$$

**Why Sigmoid?**
- Converts output into probability (0 to 1)  
- Suitable for binary classification  


## 5. Trainable Parameters Calculation

### Layer 1 (Input → Hidden)

$$
(8 \times 16) + 16 = 144
$$


### Layer 2 (Hidden → Output)

$$
(16 \times 1) + 1 = 17
$$


## Total Trainable Parameters

$$
144 + 17 = 161
$$


## Final Answer

$$
boxed{\text{Total trainable parameters} = 161}
$$


# Question 03: [ Marks 10 ]

Choose an appropriate loss function.

Choose an optimizer.

<br>

Justify your choices based on:

Regression vs Classification

Nature of the dataset

## Write Answer 03:

In [32]:
import torch.optim as optim

criterion = nn.BCELoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

BCELoss is suitable for binary classification.

Adam optimizer converges faster and adapts learning rate.

# Question 04: [ Marks 15 ]

## Implement a full training loop:

Forward pass

Loss computation

Backward pass

Parameter update

Gradient reset

### Requirements:

Train for at least 100 epochs.

Print loss every 10 epochs.

Store training loss history(You can pick your own Data Structure).

Explain clearly what happens in each step of the pipeline.

## Write Answer 04:

In [33]:
epochs = 100
loss_history = []

for epoch in range(epochs):
    outputs = model(X_train)

    loss = criterion(outputs, y_train)

    optimizer.zero_grad()
    loss.backward()

    optimizer.step()

    loss_history.append(loss.item())

    if (epoch+1) % 10 == 0:
        print(f"Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}")

Epoch [10/100], Loss: 0.7207
Epoch [20/100], Loss: 0.7052
Epoch [30/100], Loss: 0.6911
Epoch [40/100], Loss: 0.6778
Epoch [50/100], Loss: 0.6650
Epoch [60/100], Loss: 0.6524
Epoch [70/100], Loss: 0.6399
Epoch [80/100], Loss: 0.6274
Epoch [90/100], Loss: 0.6150
Epoch [100/100], Loss: 0.6026



Forward pass → prediction.

Loss computation → compare prediction with actual.

Backward pass → calculate gradients.

Optimizer step → update weights.

Zero gradients → clear old gradients.



# Question 05: [ Marks 10 ]

## Evaluate the model on test data.

## For regression:

Report MSE and MAE


## For classification:

Report Accuracy

Compare training vs testing performance.

State whether the model is underfitting or overfitting.

## Write Answer 05:

In [34]:
with torch.no_grad():
    y_pred = model(X_test)
    predicted = (y_pred > 0.5).float()

    accuracy = (predicted == y_test).float().mean()
    print("Test Accuracy:", accuracy.item()*100)

Test Accuracy: 73.3766257762909


The model is evaluated using accuracy, as it is a classification problem.

The predictions are converted into binary values using a threshold of 0.5.



If training accuracy is much higher than testing accuracy, the model is overfitting.

If both are low, the model is underfitting.

# Question 06: [ Marks 20 ]

## Modify at least ONE of the following:

Learning rate

Number of hidden neurons

Number of epochs

### Train again and compare:

Convergence speed

Final performance

Explain how the change affected the model.

## Write Answer 06:

In [35]:
epochs = 200
loss_history2 = []

for epoch in range(epochs):
    outputs = model(X_train)
    loss = criterion(outputs, y_train)

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    loss_history2.append(loss.item())

    if (epoch+1) % 20 == 0:
        print(f"Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}")

Epoch [20/200], Loss: 0.5781
Epoch [40/200], Loss: 0.5552
Epoch [60/200], Loss: 0.5343
Epoch [80/200], Loss: 0.5161
Epoch [100/200], Loss: 0.5006
Epoch [120/200], Loss: 0.4875
Epoch [140/200], Loss: 0.4767
Epoch [160/200], Loss: 0.4679
Epoch [180/200], Loss: 0.4608
Epoch [200/200], Loss: 0.4550


Modification: Increased epochs from 100 → 200

Effect:

Model learns more patterns
Loss decreases further
Accuracy improves slightly

However, too many epochs may cause overfitting.

# Question 07: [ Marks 20 ]


# Training Analysis

Answer the following:

Why must gradients be reset every epoch?

What happens if learning rate is too high?

What happens if learning rate is too small?

Why do we define layers inside the constructor (__init__) and not inside forward()?


## Write Answer 07:

1. Why reset gradients every epoch?

PyTorch accumulates gradients by default. Resetting prevents adding old gradients.

2. What happens if learning rate is too high?

Loss may oscillate or diverge.

3. What happens if learning rate is too small?

Training becomes very slow and may get stuck.

4. Why define layers inside __init__() not in forward()?

Layers need to be initialized once and their parameters tracked by PyTorch.